zict/tests/test_buffer.py

```diff
@@ -136,3 +139,76 @@ def s2f_cb(k, v):
     buff["x"]
     assert f2s == ["x", "y"]
     assert s2f == ["x"]
+
+
+def test_callbacks_exception_catch():
+    class MyError(Exception):
+        pass
+
+    f2s = []
+
+    def f2s_cb(k, v):
+        if v > 10:
+            raise MyError()
+        f2s.append(k)
+
+    s2f = []
+
+    def s2f_cb(k, v):
+        s2f.append(k)
+
+    a = dict()
+    b = dict()
+    buff = zict.Buffer(
+        a,
+        b,
+        n=10,
+        weight=lambda k, v: v,
+        fast_to_slow_callbacks=f2s_cb,
+        slow_to_fast_callbacks=s2f_cb,
+    )
+
+    buff["x"] = 1
+    buff["y"] = 2
+
+    assert buff["x"] == 1
+    assert buff["y"] == 2
+    assert not f2s
+    assert not s2f
+    assert a == {"x": 1, "y": 2}  # keys are in fast/memory
+    assert not b
+
+    # Add key < n but total weight > n this will move x out of fast
+    buff["z"] = 8
+
+    assert f2s == ["x"]
+    assert s2f == []
+    assert a == {"y": 2, "z": 8}
+    assert b == {"x": 1}
+
+    # Add key > n, again total weight > n this will move everuthing to slow except w
+    # that stays in fast due after callback raise
+    with pytest.raises(MyError):
+        buff["w"] = 11
+
+    assert f2s == ["x", "y", "z"]
+    assert s2f == []
+    assert a == {"w": 11}
+    assert b == {"x": 1, "y": 2, "z": 8}
+
+    # Retrieve x from slow to fast, will try to move w to slow and raise
+    with pytest.raises(MyError):
+        buff["x"]
+
+    assert f2s == ["x", "y", "z"]
+    # Needs FIX
+    # the following is not working because in buffer.py slow_to_fast when adding
+    # the key to fast is trying to evict "w" to slow and raising (which is an obvious
+    # fail raise we might want to avoid), not getting to the callbacks.
+    # If I do try/except and pass in slow_to_fast things are handled well, but not sure the
+    # pass is the right call here.
+    assert s2f == [
```

Comment:
Tests are broken because there is one small thing I was not quite sure how to fix. It's explained on the commented code. I wonder what's the best approach for this,

Reply:
To answer your question you should first have a legitimate use case where the buffer callback raises, and I cannot think of any. So I don't think we should try doing the right thing when we don't even know what "right" looks like.
There's also a wealth of potential issues when there is a chain of multiple callbacks and the first one fails, leaving the rest unexecuted. What does it mean? We don't know, it entirely depends on the callbacks. Again, we should not worry about it.

This whole PR is specifically to deal with ``Buffer.slow.__setitem__`` raising - let's not get off track.